packages/contracts-bedrock/contracts/universal/Proxy.sol

```diff
@@ -0,0 +1,203 @@
+    }
+
+    /**
+     * @notice Set the implementation and call a function in a single
+     *         transaction. This is useful to ensure atomic `initialize()`
+     *         based upgrades.
+     *
+     * @param _implementation The address of the implementation contract
+     * @param _data           The calldata to delegatecall the new
+     *                        implementation with
+     */
+    function setImplementationAndCall(address _implementation, bytes calldata _data)
+        external
+        payable
+        proxyCallIfNotOwner
+        returns (bytes memory)
+    {
+        _setImplementation(_implementation);
+        (bool success, bytes memory returndata) = _implementation.delegatecall(_data);
+        require(success);
+        return returndata;
+    }
+
+    /**
+     * @notice Changes the owner of the proxy contract. Only callable by the owner.
+     *
+     * @param _owner New owner of the proxy contract.
+     */
+    function setOwner(address _owner) external proxyCallIfNotOwner {
+        _setOwner(_owner);
+    }
+
+    /**
+     * @notice Gets the owner of the proxy contract.
+     *
+     * @return Owner address.
+     */
+    function getOwner() external proxyCallIfNotOwner returns (address) {
+        return _getOwner();
+    }
+
+    /**
+     * @notice Queries the implementation address.
+     *
+     * @return Implementation address.
+     */
+    function getImplementation() external proxyCallIfNotOwner returns (address) {
+        return _getImplementation();
+    }
+
+    /**
+     * @notice Sets the implementation address.
+     *
+     * @param _implementation New implementation address.
+     */
+    function _setImplementation(address _implementation) internal {
+        assembly {
+            sstore(IMPLEMENTATION_KEY, _implementation)
+        }
+        emit Upgraded(_implementation);
+    }
+
+    /**
+     * @notice Queries the implementation address.
+     *
+     * @return implementation address.
+     */
+    function _getImplementation() internal view returns (address) {
+        address implementation;
+        assembly {
+            implementation := sload(IMPLEMENTATION_KEY)
+        }
+        return implementation;
+    }
+
+    /**
+     * @notice Changes the owner of the proxy contract.
+     *
+     * @param _owner New owner of the proxy contract.
+     */
+    function _setOwner(address _owner) internal {
+        address previous = _getOwner();
+        assembly {
+            sstore(OWNER_KEY, _owner)
+        }
+        emit AdminChanged(previous, _owner);
+    }
+
+    /**
+     * @notice Queries the owner of the proxy contract.
+     *
+     * @return owner address.
+     */
+    function _getOwner() internal view returns (address) {
+        address owner;
+        assembly {
+            owner := sload(OWNER_KEY)
+        }
+        return owner;
+    }
+
+    /**
+     * @dev Performs the proxy call via a delegatecall.
+     */
+    function _doProxyCall() internal {
```

Comment:
Yes this is intentional, the call will revert anyways

Reply:
Hmmm are you sure it will revert? We should have a test for this if we don't already have that.

Reply:
Ok so it does not revert if the implementation is set to `address(0)`. Is the point of this check meant to prevent it to not call a "burn address"? Its possible to call accounts without code and non zero burn addresses